src/Microsoft.AspNetCore.Server.Kestrel.Transport.Windows.Rio/Internal/Networking/RioCompletionThread.cs

```diff
@@ -0,0 +1,384 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.Buffers;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Windows.Rio.Internal.Networking
+{
+    public class RioCompletionThread : RioThreadContext
+    {
+        private const int MaxSocketsPerThread = 256000;
+        private const int MaxReadsPerSocket = 1;
+        public const int MaxWritesPerSocket = 2;
+        public const int MaxOutsandingCompletionsPerThread = (MaxReadsPerSocket + MaxWritesPerSocket) * MaxSocketsPerThread;
+
+        private readonly Thread _thread;
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly object _startSync = new object();
+        private bool _initCompleted = false;
+
+        private readonly TaskCompletionSource<int> _threadTcs = new TaskCompletionSource<int>();
+
+        private RioCompletionPort _completionPort;
+        private RioCompletionQueue _completionQueue;
+
+        private Dictionary<long, RioConnection> _connections;
+        private List<BufferMapping> _bufferIdMappings;
+
+        public RioCompletionPort CompletionPort => _completionPort;
+        public RioCompletionQueue CompletionQueue => _completionQueue;
+
+        public bool NotCancelled => !_cancellationToken.IsCancellationRequested;
+
+        public long _connectionsCreated;
+
+        bool _activatedNotify;
+
+        public RioCompletionThread(RioListenerContext context, int threadId, CancellationToken cancellationToken) : base(context)
+        {
+            _cancellationToken = cancellationToken;
+
+            _thread = new Thread(Run)
+            {
+                Name = $"RIO Completion Thread {threadId:00}",
+                IsBackground = true
+            };
+        }
+
+        public Task StartAsync()
+        {
+            var tcs = new TaskCompletionSource<int>();
+            _thread.Start(
+                new RioThreadStart()
+                {
+                    State = this,
+                    StartTcs = tcs
+                });
+
+            return tcs.Task;
+        }
+
+        public async Task StopAsync(TimeSpan timeout)
+        {
+            await Task.Delay(0);
+            lock (_startSync)
+            {
+                if (!_initCompleted)
+                {
+                    return;
+                }
+            }
+
+            if (!_threadTcs.Task.IsCompleted)
+            {
+                DisposeConnections();
+            }
+        }
+
+        private void DisposeConnections()
+        {
+            //try
```

Comment:
What is this dead code 😄 . @halter73 is working on moving shutdown to kestrel core so you won't have to handle this.